File: visible-ios/visible/visible/NodeDetailLogic.swift
import Foundation
import os.log

private let logger = Logger.visible("NodeDetailLogic")

/// The pure form conversions behind ``NodeDetailModel``: the value cents↔dollars
/// pair, the acquired-date ISO↔`Date` pair, and the blank/quantity parsing the
/// form seeds from and saves back through. No `AppHandle` or `@Observable`
/// state, so the round trips and the unparseable→nil cases are exercised
/// directly. The model delegates each conversion here.
enum NodeDetailLogic {
    /// The ISO `YYYY-MM-DD` formatter the acquired date seeds from and saves back
    /// to. Fixed locale and UTC so a bare calendar date round-trips without a
    /// time-zone shift.
    private static let isoDate: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    /// Parse a stored ISO `YYYY-MM-DD` string into a `Date?` for the native
    /// picker (form-seeding); a string that doesn't parse seeds no date — logged,
    /// like the Android counterpart and the value/quantity parses here.
    static func dateFromIso(_ iso: String) -> Date? {
        if let date = isoDate.date(from: iso) { return date }
        logger.debug("acquired date \(iso, privacy: .public) is not an ISO date; showing no date")
        return nil
    }

    /// Render a picked `Date` back to the stored ISO `YYYY-MM-DD` string.
    static func isoFromDate(_ date: Date) -> String {
        isoDate.string(from: date)
    }

    /// A trimmed text field, or nil when blank — the absence a cleared field means.
    static func blankToNil(_ text: String) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    /// Parse a whole-number string into `Int64?`. Blank is nil (a cleared field,
    /// the form-seeding exemption). A non-blank string that doesn't parse is also
    /// nil, but that's a dropped value on the save path, so it's logged.
    static func int64(from text: String) -> Int64? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        guard let value = Int64(trimmed) else {
            logger.warning("quantity \(trimmed, privacy: .public) is not a whole number; saving no quantity")
            return nil
        }
        return value
    }

    /// Cents → a dollars string with two decimal places (form-seeding).
    static func dollars(fromCents cents: Int64) -> String {
        let value = Decimal(cents) / 100
        return NSDecimalNumber(decimal: value).stringValue
    }

    /// A dollars string → whole cents, rounded to the nearest cent; blank or
    /// unparseable is nil. Uses `Decimal` so the cent rounding doesn't inherit
    /// binary-float drift.
    static func cents(fromDollars text: String) -> Int64? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        guard let dollars = Decimal(string: trimmed, locale: Locale(identifier: "en_US_POSIX")) else {
            logger.warning("value \(trimmed, privacy: .public) is not a parseable amount; saving no value")
            return nil
        }
        var cents = dollars * 100
        var rounded = Decimal()
        NSDecimalRound(&rounded, &cents, 0, .plain)
        return NSDecimalNumber(decimal: rounded).int64Value
    }
}

File: visible-android/app/src/main/java/fm/bae/visible/ui/NodeDetailLogic.kt
package fm.bae.visible.ui

import android.util.Log
import java.math.BigDecimal
import java.math.RoundingMode
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

private const val TAG = "visible.NodeDetailLogic"

/**
 * The pure form conversions behind [NodeDetailViewModel]: the value cents↔dollars
 * pair, the acquired-date ISO↔UTC-epoch-millis pair (the unit the Material date
 * picker speaks), and the blank/quantity parsing the form seeds from and saves
 * back through. No [uniffi.visible_bridge.AppHandle] or observable state, so the
 * round trips and the unparseable→null cases are exercised directly. The view
 * model delegates each conversion here.
 */
object NodeDetailLogic {
    private val ISO_DATE: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

    /**
     * A quantity string → [Long], or null. Blank is null (a cleared field, the
     * form-seeding exemption). A non-blank string that isn't a whole number is
     * also null, but that's a dropped value on the save path, so it's logged.
     */
    fun quantityFromText(text: String): Long? {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return null
        val value = trimmed.toLongOrNull()
        if (value == null) {
            Log.d(TAG, "quantity '$trimmed' is not a whole number; saving no quantity")
        }
        return value
    }

    /** Cents → a dollars string with two decimal places (form-seeding). */
    fun dollarsFromCents(cents: Long): String =
        BigDecimal(cents).movePointLeft(2).toPlainString()

    /**
     * A dollars string → whole cents, rounded to the nearest cent; blank or
     * unparseable is null. Uses [BigDecimal] so the cent rounding doesn't inherit
     * binary-float drift.
     */
    fun centsFromDollars(text: String): Long? {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return null
        return try {
            BigDecimal(trimmed).movePointRight(2).setScale(0, RoundingMode.HALF_UP).longValueExact()
        } catch (e: ArithmeticException) {
            Log.d(TAG, "value '$trimmed' is not a parseable amount; saving no value", e)
            null
        } catch (e: NumberFormatException) {
            Log.d(TAG, "value '$trimmed' is not a parseable amount; saving no value", e)
            null
        }
    }

    /** An ISO `YYYY-MM-DD` date → UTC start-of-day epoch millis (form-seeding). */
    fun millisFromIso(iso: String): Long? =
        try {
            LocalDate.parse(iso, ISO_DATE).atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli()
        } catch (e: DateTimeParseException) {
            Log.d(TAG, "acquired date '$iso' is not an ISO date; showing no date", e)
            null
        }

    /** UTC epoch millis → an ISO `YYYY-MM-DD` date string. */
    fun isoFromMillis(millis: Long): String =
        Instant.ofEpochMilli(millis).atZone(ZoneOffset.UTC).toLocalDate().format(ISO_DATE)
}
